File: frontend/lib/api/p115.ts
import { authFetch } from './client'

export interface P115Status {
  logged_in: boolean
  app?: string
  updated_at?: string
}

export interface P115QrcodeToken {
  uid: string
  time: number
  sign: string
  qrcode_content: string
}

export interface P115PollResult {
  status: number
  msg: string
}

/** 115 容量项（size + size_format） */
export interface P115SizeItem {
  size: number
  size_format: string
}

/** 115 空间信息 */
export interface P115SpaceInfo {
  all_total?: P115SizeItem | null
  all_remain?: P115SizeItem | null
  all_use?: P115SizeItem | null
}

/** 115 存储信息（fs_index_info data） */
export interface P115StorageInfo {
  space_info?: P115SpaceInfo | null
}

/** 115 VIP 信息 */
export interface P115VipInfo {
  is_vip?: boolean
  is_forever?: boolean
  expire?: string
  expire_str?: string
}

/** 115 头像 */
export interface P115FaceInfo {
  face_l?: string
  face_m?: string
  face_s?: string
}

/** 115 用户信息（user_my_info data） */
export interface P115UserInfo {
  uid: number
  uname?: string
  vip?: P115VipInfo | null
  face?: P115FaceInfo | null
}

/** 115 仪表盘数据 */
export interface P115Dashboard {
  logged_in: boolean
  user_info: P115UserInfo | null
  storage_info: P115StorageInfo | null
}

/**
 * 获取 115 仪表盘数据（用户信息 + 存储信息），未登入时 user_info、storage_info 为 null
 */
export async function apiP115Dashboard(token: string): Promise<P115Dashboard> {
  const res = await authFetch(token, '/api/v1/p115/dashboard')
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || '获取仪表盘失败')
  }
  return res.json()
}

/**
 * 获取 115 登入状态
 */
export async function apiP115Status(token: string): Promise<P115Status> {
  const res = await authFetch(token, '/api/v1/p115/status')
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || '获取状态失败')
  }
  return res.json()
}

/**
 * 获取二维码 token
 */
export async function apiP115QrcodeToken(token: string, app: string = 'qandroid'): Promise<P115QrcodeToken> {
  const res = await authFetch(token, `/api/v1/p115/qrcode/token?app=${encodeURIComponent(app)}`, {
    method: 'POST',
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || '获取二维码失败')
  }
  return res.json()
}

/**
 * 获取二维码图片（作为 blob URL）
 */
export async function apiP115QrcodeImage(token: string, uid: string): Promise<string> {
  const res = await authFetch(token, `/api/v1/p115/qrcode/image?uid=${encodeURIComponent(uid)}`, {
    headers: { Accept: 'image/png' },
  })
  if (!res.ok) {
    throw new Error('获取二维码图片失败')
  }
  const blob = await res.blob()
  return URL.createObjectURL(blob)
}

/**
 * 轮询扫码状态
 */
export async function apiP115QrcodePoll(
  token: string,
  uid: string,
  time: number,
  sign: string
): Promise<P115PollResult> {
  const params = new URLSearchParams({ uid, time: String(time), sign })
  const res = await authFetch(token, `/api/v1/p115/qrcode/poll?${params}`)
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || '轮询失败')
  }
  return res.json()
}

/**
 * 确认扫码登入
 */
export async function apiP115QrcodeConfirm(
  token: string,
  uid: string,
  app: string = 'qandroid'
): Promise<{ ok: boolean; cookies_str: string }> {
  const params = new URLSearchParams({ uid, app })
  const res = await authFetch(token, `/api/v1/p115/qrcode/confirm?${params}`, {
    method: 'POST',
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || '确认登入失败')
  }
  return res.json()
}

/**
 * 退出 115 登录
 */
export async function apiP115Logout(token: string): Promise<{ ok: boolean }> {
  const res = await authFetch(token, '/api/v1/p115/logout', { method: 'POST' })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || '退出失败')
  }
  return res.json()
}

File: backend/app/schemas/p115.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class P115StatusResponse(BaseModel):
    """
    115 登入状态响应
    """

    logged_in: bool = Field(..., description="是否已登入")
    app: str | None = Field(default=None, description="登入设备/应用")
    updated_at: datetime | None = Field(default=None, description="登入时间（UTC）")


class P115QrcodeTokenResponse(BaseModel):
    """
    115 二维码 token 响应
    """

    uid: str = Field(..., description="二维码 uid")
    time: int = Field(..., description="时间戳")
    sign: str = Field(..., description="签名")
    qrcode_content: str = Field(..., description="二维码内容")


class P115PollResponse(BaseModel):
    """
    115 扫码轮询状态响应
    """

    status: int = Field(..., description="扫码状态码")
    msg: str = Field(..., description="状态说明")


class P115ConfirmResponse(BaseModel):
    """
    115 确认扫码响应
    """

    ok: bool = Field(..., description="是否成功")
    cookies_str: str = Field(..., description="保存的 cookies 字符串")


class P115LogoutResponse(BaseModel):
    """
    115 退出登录响应
    """

    ok: bool = Field(..., description="是否成功")


class P115VipInfo(BaseModel):
    """
    115 VIP 信息
    """

    is_vip: bool = Field(default=False, description="是否 VIP")
    is_forever: bool = Field(default=False, description="是否永久")
    expire: str = Field(default="", description="过期时间戳")
    expire_str: str = Field(default="", description="过期时间描述，如 2026-07-30")


class P115FaceInfo(BaseModel):
    """
    115 头像
    """

    face_l: str = Field(default="", description="大图")
    face_m: str = Field(default="", description="中图")
    face_s: str = Field(default="", description="小图")


class P115UserInfo(BaseModel):
    """
    115 用户信息
    """

    model_config = ConfigDict(extra="allow")

    uid: int = Field(..., description="用户 ID")
    uname: str = Field(default="", description="用户名")
    vip: P115VipInfo | None = Field(default=None, description="VIP 信息")
    face: P115FaceInfo | None = Field(default=None, description="头像")


class P115SizeItem(BaseModel):
    """
    容量项（size + size_format）
    """

    size: int = Field(default=0, description="字节数")
    size_format: str = Field(default="", description="如 21.01TB")


class P115SpaceInfo(BaseModel):
    """
    115 空间信息（space_info）
    """

    all_total: P115SizeItem | None = Field(default=None, description="总容量")
    all_remain: P115SizeItem | None = Field(default=None, description="剩余容量")
    all_use: P115SizeItem | None = Field(default=None, description="已用容量")


class P115StorageInfo(BaseModel):
    """
    115 存储/索引信息
    """

    model_config = ConfigDict(extra="allow")

    space_info: P115SpaceInfo | None = Field(default=None, description="空间信息")


class P115DashboardResponse(BaseModel):
    """
    115 仪表盘数据（用户信息 + 存储信息）
    """

    logged_in: bool = Field(..., description="是否已登入 115")
    user_info: P115UserInfo | None = Field(
        default=None, description="115 用户信息，未登入为 None"
    )
    storage_info: P115StorageInfo | None = Field(
        default=None, description="115 存储/索引信息，未登入为 None"
    )
